src/lib/device-state.ts

```diff
@@ -97,6 +97,33 @@ export const serviceInstallFromImage = (
 export const formatImageLocation = (imageLocation: string) =>
 	imageLocation.toLowerCase();
 
+export type DeviceConfigHook = (
+	configVars: Dictionary<string>,
+	osVersion: string,
+) => void;
+
+export interface HookRegistry {
+	deviceConfig: DeviceConfigHook[];
+}
+
+const STATE_ENDPOINT_HOOKS: HookRegistry = {
+	deviceConfig: [],
+};
+
+export function registerHook<
+	Key extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[Key][number]
+>(name: Key, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook<
+	Key extends keyof HookRegistry,
+	Args extends Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>
```

Comment:
This allows changing what `Args` is, which shouldn't be possible as it should always be precisely `Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>`, hence why I used `args: Parameters<HookRegistry[T][number]>`